library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java

```diff
@@ -379,10 +379,9 @@ private void feedDataToExtractor(
         while (result == Extractor.RESULT_CONTINUE && !loadCanceled) {
           result = extractor.read(input, DUMMY_POSITION_HOLDER);
         }
-      } catch (UnreportedDiscontinuityException e) {
-        Log.d(TAG, "Unreported discontinuity at timeUs: "+ e.timesUs + " uri: " + dataSpec.uri);
-        throw new IOException("Timestamp error", e);
-        
+      } catch (UnexpectedDiscontinuityException e) {
+        Log.d(TAG, "UnexpectedDiscontinuityException - recovering by discarding balance of segment", e);
+        throw new IOException("load aborted for segment - " + e.dataSpec + " unexpected discontinuity", e);
       } finally {
```

Comment:
`UnexpectedDiscontinuityException` needs to stay a `RuntimeException` until signatures in the class hierarchy for `TrackOutput#sampleMetadata()` are changed to include a declared exception.  Best to keep this in a separate change to avoid having to merge/rebase lots of files while this branch is open.

Reply:
Can you just make it an `IOException`, to avoid needing to catch and re-throw?

Reply:
I still need to catch in order to get to the timestamp adjuster (and/or add metadata to the segment to discard the time period and recover).  I can probably do this on my own fork if no one else would want it (my preference always is ExoPlayer proper to take the changes, note if by default we fail playback anything done here to facilitate recovery will not matter).

This latest checkin (4701a6b39081cccaab202effb6cf59de6d4510b2) make the exception extend IOException (much nicer, I agree, then a runtime exception).   It took a while to make and run the tests, so hopefully I'll not have to rebase it much (Most of it is of course pretty rote to review and in the extractor library)